### DISCUSSION TITLE
How to early-stop optimization while multi objective ?

### DISCUSSION THREAD

**Discussion by eromoe** [optuna/optuna#4492] [Q&A]
I'd like early stop optimization in case no improvement within 10 iter either objective score . 

My alter function is like 
```python
class MultiObjectiveEarlyStoppingCallback(object):
    """Early stopping callback for Optuna."""

    def __init__(self, early_stopping_rounds: int, directions:Union[str, List[str]] = "minimize") -> None:
        self.early_stopping_rounds = early_stopping_rounds

        self._iter = 0
        self._operator = []
        self._score = []
        
        if isinstance(directions, str):
            directions = [directions]
            
        for direction in directions:
            if direction == "minimize":
                self._operator.append(operator.lt)
                self._score.append(np.inf)
            elif direction == "maximize":
                self._operator.append(operator.gt)
                self._score.append(-np.inf)
            else:
                ValueError(f"invalid direction: {direction}")

    def __call__(self, study: optuna.Study, trial: optuna.Trial) -> None:
        """Do early stopping."""
        for i in range(len(self._operator)):
            if self._operator[i](study.best_values[i], self._score[i]):
                self._iter = 0
                self._score[i] = study.best_values[i]
            else:
                self._iter += 1

            if self._iter >= self.early_stopping_rounds:
                study.stop()
```

But `study.best_values` doesn't exists , make me confusing .

**Comment by nzw0301** [comment]
Could you share minimal reproducible and executable codes with us?

**Comment by eromoe** [comment]
That code is altered from https://github.com/optuna/optuna/issues/1001#issuecomment-862843041
, can't run in fact .  I write it for an exmaple to point out the problem.
As you see,  study with multi objective may have different directions,  and there is no an obvious way to retrieve corresponding best_value of each objective .

**Reply by nzw0301** [reply]
Since you access an undefined variable of the study, the error isn't surprising. I'm not sure how to perform early stopping with multi-objective value since the best values are on the Pareto-front line, not a single point.

I highly recommend reading multi-objective tutorial: https://optuna.readthedocs.io/en/stable/tutorial/20_recipes/002_multi_objective.html.
